### PR TITLE
implements #677

### PR DESCRIPTION
**Pull request by dachziegel**

### CODE REVIEW

File: thrust/system/detail/generic/unique.inl [hunk 1]
```diff
@@ -101,13 +102,11 @@ __host__ __device__
   if(first == last)
```
Comment (jaredhoberock) [inline thrust/system/detail/generic/unique.inl]: i bet this check is not necessary with the new implementation

File: thrust/system/detail/generic/unique_by_key.inl [hunk 2]
```diff
@@ -108,26 +109,23 @@ template<typename ExecutionPolicy,
                        BinaryPredicate binary_pred)
 {
   typedef typename thrust::iterator_traits<InputIterator1>::difference_type difference_type;
-  
+
   // empty sequence
   if(keys_first == keys_last)
     return thrust::make_pair(keys_output, values_output);
```
Comment (jaredhoberock) [inline thrust/system/detail/generic/unique_by_key.inl]: same here

**Comment by jaredhoberock** [comment]
New implementation looks great. I'd see if those checks for empty sequences are no longer necessary because `copy_if` should be able to handle empty input.

Do these changes pass unit tests? I'll test them on Windows locally.

**Comment by dachziegel** [comment]
Yes, unit tests are passed on Ubuntu 14.04 & CUDA 7.

**Comment by dachziegel** [comment]
The latest commit  (28eed1828d520632f80ce708f399c7877c0393b4) also passed the unit tests.

**Comment by jaredhoberock** [comment]
It passes on Windows. One last thing which I forgot before -- can you add an entry for this enhancement in the CHANGELOG as well as an acknowledgement? I'll merge at that point.

**Comment by dachziegel** [comment]
done

**Comment by jaredhoberock** [comment]
Thanks, great work!

**Pull request merged**